File: 2023/day-01/c/part2.c
#include <ctype.h>
#include <stdio.h>
#include <string.h>

int isstrnum(char *p);

int main(void) {
  char *line = NULL;
  size_t len = 0;
  int sum = 0;

  while (getline(&line, &len, stdin) != -1) {
    char *start = line;
    char *end = start + strlen(line);

    while (!(isdigit(*start) || isstrnum(start))) {
      start++;
    }

    while (!(isdigit(*end) || isstrnum(end))) {
      end--;
    }

    sum += (*start - '0') * 10 + (*end - '0');
  }

  printf("%d", sum);
}

int isstrnum(char *p) {
#define N 10

  char *ns[N] = {"zero", "one", "two",   "three", "four",
                 "five", "six", "seven", "eight", "nine"};

  for (int i = 0; i < N; i++) {
    if (strncmp(ns[i], p, strlen(ns[i])) == 0) {
      *p = i + '0';
      return 1;
    }
  }

  return 0;
}

File: 2023/day-01/javascript/part2.js
const numbers = {
  one: 1,
  two: 2,
  three: 3,
  four: 4,
  five: 5,
  six: 6,
  seven: 7,
  eight: 8,
  nine: 9,
};

export default (input) =>
  input
    .split("\n")
    .filter(Boolean)
    .map((line) => {
      for (const [strNum, num] of Object.entries(numbers)) {
        line = line.replaceAll(
          strNum,
          `${strNum.slice(0, 1)}${num}${strNum.slice(1)}`,
        );
      }
      const nums = line.replaceAll(/\D/g, "");
      return nums[0] + nums[nums.length - 1];
    })
    .reduce((acc, cur) => acc + Number(cur), 0);
